core/network/portrange_test.go

```diff
@@ -428,3 +428,70 @@ func (p *PortRangeSuite) TestUniquePortRangesInGroup(c *gc.C) {
 	got := in.UniquePortRanges()
 	c.Assert(got, gc.DeepEquals, exp, gc.Commentf("expected duplicate port ranges to be removed"))
 }
+
+func (p *PortRangeSuite) TestGroupedPortRangesEquality(c *gc.C) {
+	specs := []struct {
+		descr    string
+		a, b     network.GroupedPortRanges
+		expEqual bool
+	}{
+		{
+			descr: "equal port ranges in random order",
+			a: network.GroupedPortRanges{
+				"foo": []network.PortRange{
+					network.MustParsePortRange("123/tcp"),
+					network.MustParsePortRange("456/tcp"),
+				},
+				"bar": []network.PortRange{
+					network.MustParsePortRange("123/tcp"),
+				},
+			},
+			b: network.GroupedPortRanges{
+				"foo": []network.PortRange{
+					network.MustParsePortRange("456/tcp"),
+					network.MustParsePortRange("123/tcp"),
+				},
+				"bar": []network.PortRange{
+					network.MustParsePortRange("123/tcp"),
+				},
+			},
+			expEqual: true,
+		},
+		{
+			descr: "groups with different lengths",
+			a: network.GroupedPortRanges{
+				"foo": []network.PortRange{
+					network.MustParsePortRange("123/tcp"),
+					network.MustParsePortRange("456/tcp"),
+				},
+			},
+			b: network.GroupedPortRanges{
+				"foo": []network.PortRange{
+					network.MustParsePortRange("123/tcp"),
+				},
+			},
+			expEqual: false,
+		},
+		{
+			descr: "groups with same length but different keys",
+			a: network.GroupedPortRanges{
+				"foo": []network.PortRange{
+					network.MustParsePortRange("123/tcp"),
+					network.MustParsePortRange("456/tcp"),
+				},
+			},
+			b: network.GroupedPortRanges{
+				"bar": []network.PortRange{
+					network.MustParsePortRange("123/tcp"),
+				},
```

Comment:
but these aren't the same length (according to descr)

I don't know if the code is incorrect, but the test data doesn't match the description.

Reply:
The description refers to the length of the GroupedPortRanges _keys_ and not their values. It is meant to test the slow path (i.e. they have same number of keys so we need to iterate and compare both keys and their contents). I can update the description if it seems confusing.